plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/purity/PurityChecker.kt

```diff
@@ -0,0 +1,60 @@
+package org.jetbrains.kotlin.formver.purity
+
+import org.jetbrains.kotlin.formver.embeddings.callables.UserFunctionEmbedding
+import org.jetbrains.kotlin.formver.embeddings.expression.*
+
+class PurityChecker() {
+
+    // Keeping these because they might be useful later?
+    /*
+    private var debugSink: ((String) -> Unit)? = null
+    fun setDebugSink(sink: ((String) -> Unit)?) {
+        debugSink = sink
+    }
+
+    private inline fun dbg(msg: () -> String) {
+        debugSink?.invoke(msg())
+    }
+    */
+
+    /**
+     * We restrict our selves to using a helper function for now that only evaluates asserts.
+     */
+    fun checkIsPure(userFunction: UserFunctionEmbedding) {
+        val exp = userFunction.body ?: return
+        assertsArePure(exp.debugExpEmbedding!!)
+    }
+
+    private fun isExpressionPure(exp: ExpEmbedding): Boolean = when (exp) {
+        is PureExpEmbedding -> true
+        is Block -> exp.exps.all(::isExpressionPure)
+
+        // We only check the right hand side, since for asserts, the left hand side will always be PlaceholderVariableEmbedding (which is pure)
+        is Assign -> isExpressionPure(exp.rhs)
+        is PassthroughExpEmbedding -> isExpressionPure(exp.inner)
+        is BinaryDirectResultExpEmbedding ->
+            isExpressionPure(exp.left) && isExpressionPure(exp.right)
+        is UnaryDirectResultExpEmbedding ->
+            isExpressionPure(exp.inner)
+        // Here we will check whether the function is part of our special functions but should also check if the function is annotated as pure if it is not whitelisted
+        is MethodCall -> {
+            true
+        }
+        is If -> isExpressionPure(exp.condition) &&
+                isExpressionPure(exp.thenBranch) &&
+                isExpressionPure(exp.elseBranch)
+        else -> false
+    }
+
+    private fun assertsArePure(exp: ExpEmbedding): Boolean = when (exp) {
+        is Assert -> isExpressionPure(exp.exp)
```

Comment:
This also permits cases like `verify(verify(x))` which we probably don't want to allow.